src/oidcendpoint/oauth2/authorization.py

```diff
@@ -100,27 +287,51 @@ class Authorization(Endpoint):
 
     def __init__(self, endpoint_context, **kwargs):
         Endpoint.__init__(self, endpoint_context, **kwargs)
-        # self.pre_construct.append(self._pre_construct)
         self.post_parse_request.append(self._do_request_uri)
         self.post_parse_request.append(self._post_parse_request)
         self.allowed_request_algorithms = AllowedAlgorithms(ALG_PARAMS)
-        # Has to be done elsewhere. To make sure things happen in order.
-        # self.scopes_supported = available_scopes(endpoint_context)
 
     def filter_request(self, endpoint_context, req):
         return req
 
+    def extra_response_args(self, aresp):
+        return aresp
+
     def verify_response_type(self, request, cinfo):
         # Checking response types
         _registered = [set(rt.split(" ")) for rt in cinfo.get("response_types", [])]
         if not _registered:
             # If no response_type is registered by the client then we'll
-            # code which it the default according to the OIDC spec.
+            # use code.
             _registered = [{"code"}]
 
         # Is the asked for response_type among those that are permitted
         return set(request["response_type"]) in _registered
 
+    def mint_token(self, token_type, grant, session_id, based_on=None):
+        _mngr = self.endpoint_context.session_manager
+        usage_rules = grant.usage_rules.get(token_type, {})
+
+        token = grant.mint_token(
+            session_id=session_id,
+            endpoint_context=self.endpoint_context,
+            token_type=token_type,
+            token_handler=_mngr.token_handler["access_token"],
+            based_on=based_on,
+            usage_rules=usage_rules
+        )
+
+        _exp_in = usage_rules.get("expires_in")
+        if isinstance(_exp_in, str):
+            _exp_in = int(_exp_in)
+        if _exp_in:
+            token.expires_at = utc_time_sans_frac() + _exp_in
```

Comment:
Shouldn't this be done in `grant.mint_token`?

Reply:
mint_token() only sets expires_at and that is because expires_at is relative to the minter's time and is absolute.
When a token is presented to the introspection endpoint, expires_at is used to see if the token has timed-out.
One could also use expires_in + issued_at but I choose to go this way.

expires_in on the other hand is relative which means you have to know when it was minted or when you received it to be able to figure out if the token is too old.